lib/src/annotate.rs

```diff
@@ -243,9 +265,28 @@ impl Source {
     }
 }
 
-/// List of commit IDs that originated lines, indexed by line numbers in the
-/// original file.
-type OriginalLineMap = Vec<Result<CommitId, CommitId>>;
+/// List of sources that originated lines, indexed by line numbers in the
+/// starting file.
+type OriginalLineMap = Vec<Result<LineSource, CommitId>>;
```

Comment:
nit: simply make it store `Result<LineSource, LineSource>`? It's a bit odd that the intermediate value lacks line number whereas the final value knows one.

Reply:
That was my original plan, but it turns out that this information wasn't accurate enough in my tests. In this case:

https://github.com/jj-vcs/jj/blob/2747dafeb3390bc7a939571f759f3c5184f5a147/lib/src/annotate.rs#L380-L388

The current line number here (first value in the tuple, `_`) doesn't always match the number I expected it to be, and I couldn't figure out another way to obtain it.

Reply:
So, the issue isn't so much intermediate values, but the final state when the source possibly falls outside of the domain.

Reply:
I've pushed this change experimentally here, for comparison:

- Changing to `Err(parent_commit_id)` in isolation: https://github.com/mernen/jj/commit/3d1d83f75e67d89344eb5b1d5fa3bef579077844
- This PR on top of it: https://github.com/jj-vcs/jj/compare/main...mernen:jj:push-rqzwpuzqrpzm

Observations:

- Line numbers are indeed correctly tracked when we do this, so this fork uses `Result<LineOrigin, LineOrigin>`
- This appears to be closer to what `git blame` does
- This isn't currently represented in the tests, but I found the results somewhat more accurate when there are gaps in the domain

Personally, I'm in favor of changing the behavior for the error case.